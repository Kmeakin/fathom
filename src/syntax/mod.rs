//! The syntax of the language

pub mod concrete;
pub mod core;
pub mod parse;
pub mod pretty;
pub mod translation;
