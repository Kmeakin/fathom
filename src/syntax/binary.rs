//! The syntax of our data description language

use source::Span;
use syntax::{self, host, Field, Named, Var};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Kind {
    /// Kind of types
    Type,
    /// Kind of type functions
    Arrow(Box<Kind>, Box<Kind>),
}

impl Kind {
    /// Kind of type functions
    pub fn arrow<K1: Into<Box<Kind>>, K2: Into<Box<Kind>>>(lhs: K1, rhs: K2) -> Kind {
        Kind::Arrow(lhs.into(), rhs.into())
    }
}

/// A binary type
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TypeF<N, T, E> {
    /// A type variable: eg. `T`
    Var(Var<N, Named<N, u32>>),
    /// Bit
    Bit,
    /// An array of the specified type, with a size: eg. `[T; n]`
    Array(Box<T>, Box<E>),
    /// A union of types: eg. `union { T, ... }`
    Union(Vec<T>),
    /// A struct type, with fields: eg. `struct { field : T, ... }`
    Struct(Vec<Field<N, T>>),
    /// A type constrained by a predicate: eg. `T where x => x == 3`
    Cond(Named<N, Box<T>>, Box<E>),
    /// Type abstraction: eg. `\(a : Type) -> T`
    Abs(Named<N, Kind>, Box<T>),
    /// Type application: eg. `T U V`
    App(Box<T>, Box<T>),
}

/// A binary type AST
pub trait TypeNode<N, E: host::ExprNode<N>>
    : Clone + AsRef<TypeF<N, Self, E>> + AsMut<TypeF<N, Self, E>> {
}

/// The recursive innards of a `Type`
///
/// This does the job of tying the recursive knot for `Type`, turning
/// `TypeF` into a tree of expressions.
///
/// We could have just inlined this inside `Type`, but this type alias is
/// handy to have around for defining our conversion impls.
pub type TypeRec<N, E> = TypeF<N, Type<N, E>, E>;

/// A tree of binary types
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Type<N, E>(pub TypeRec<N, E>);

impl<N: Clone, E: host::ExprNode<N>> TypeNode<N, E> for Type<N, E> {}

impl<N, E> Into<TypeRec<N, E>> for Type<N, E> {
    fn into(self) -> TypeRec<N, E> {
        self.0
    }
}

impl<N, E> AsRef<TypeRec<N, E>> for Type<N, E> {
    fn as_ref(&self) -> &TypeRec<N, E> {
        &self.0
    }
}

impl<N, E> AsMut<TypeRec<N, E>> for Type<N, E> {
    fn as_mut(&mut self) -> &mut TypeRec<N, E> {
        &mut self.0
    }
}

/// The recursive innards of a `SpannedType`
///
/// This does the job of tying the recursive knot for `SpannedType`, turning
/// `TypeF` into a tree of spanned expressions.
///
/// We could have just inlined this inside `SpannedType`, but this type alias is
/// handy to have around for defining our conversion impls.
pub type SpannedTypeRec<N, E> = TypeF<N, SpannedType<N, E>, E>;

/// A tree of spanned binary types
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SpannedType<N, E> {
    pub span: Span,
    pub inner: SpannedTypeRec<N, E>,
}

impl<N: Clone, E: host::ExprNode<N>> TypeNode<N, E> for SpannedType<N, E> {}

impl<N, E> From<TypeRec<N, E>> for Type<N, E> {
    fn from(src: TypeRec<N, E>) -> Type<N, E> {
        Type(src)
    }
}

impl<N, E> AsRef<SpannedTypeRec<N, E>> for SpannedType<N, E> {
    fn as_ref(&self) -> &SpannedTypeRec<N, E> {
        &self.inner
    }
}

impl<N, E> AsMut<SpannedTypeRec<N, E>> for SpannedType<N, E> {
    fn as_mut(&mut self) -> &mut SpannedTypeRec<N, E> {
        &mut self.inner
    }
}

impl<N, T, E> TypeF<N, T, E> {
    /// A free type variable: eg. `T`
    pub fn fvar<N1: Into<N>>(x: N1) -> TypeF<N, T, E> {
        TypeF::Var(Var::Free(x.into()))
    }

    /// A bound type variable
    pub fn bvar<N1: Into<N>>(x: N1, i: u32) -> TypeF<N, T, E> {
        TypeF::Var(Var::Bound(Named(x.into(), i)))
    }

    /// An array of the specified type, with a size: eg. `[T; n]`
    pub fn array<T1: Into<Box<T>>, E1: Into<Box<E>>>(elem_ty: T1, size_expr: E1) -> TypeF<N, T, E> {
        TypeF::Array(elem_ty.into(), size_expr.into())
    }

    /// A union of types: eg. `union { T, ... }`
    pub fn union(tys: Vec<T>) -> TypeF<N, T, E> {
        TypeF::Union(tys)
    }

    /// Type application: eg. `T U V`
    pub fn app<T1: Into<Box<T>>, T2: Into<Box<T>>>(ty1: T1, ty2: T2) -> TypeF<N, T, E> {
        TypeF::App(ty1.into(), ty2.into())
    }
}

impl<N, T, E> TypeF<N, T, E>
where
    T: TypeNode<N, E>,
    E: host::ExprNode<N>,
{
    /// A struct type, with fields: eg. `struct { field : T, ... }`
    pub fn struct_<Fs>(fields: Fs) -> TypeF<N, T, E>
    where
        N: PartialEq + Clone,
        Fs: IntoIterator<Item = Field<N, T>>,
    {
        // We maintain a list of the seen field names. This will allow us to
        // recover the index of these variables as we abstract later fields...
        let mut seen_names = Vec::new();
        let f = |field: Field<N, T>| {
            let field = field.map_value(|mut ty| {
                T::as_mut(&mut ty).abstract_with(&|x| {
                    seen_names
                        .iter()
                        .position(|y| x == y)
                        .map(|i| Named(x.clone(), i as u32))
                });
                ty
            });

            // Record that the field has been 'seen'
            seen_names.push(field.name.clone());

            field
        };

        TypeF::Struct(fields.into_iter().map(f).collect())
    }

    /// A type constrained by a predicate: eg. `T where x => x == 3`
    pub fn cond<N1, T1, E1>(ty: T1, param: N1, pred: E1) -> TypeF<N, T, E>
    where
        N: PartialEq + Clone,
        N1: Into<N>,
        T1: Into<Box<T>>,
        E1: Into<Box<E>>,
    {
        let param = param.into();
        let mut pred = pred.into();
        E::as_mut(&mut pred).abstract_name(&param);

        TypeF::Cond(Named(param, ty.into()), pred)
    }

    /// Type abstraction: eg. `\(a : Type) -> T`
    pub fn abs<T1>(param: Named<N, Kind>, body_ty: T1) -> TypeF<N, T, E>
    where
        N: PartialEq + Clone,
        T1: Into<Box<T>>,
    {
        let mut body_ty = body_ty.into();
        T::as_mut(&mut body_ty).abstract_name(&param.0);
        TypeF::Abs(param, body_ty)
    }

    fn abstract_level_with<F>(&mut self, level: u32, f: &F)
    where
        F: Fn(&N) -> Option<Named<N, u32>>,
    {
        match *self {
            TypeF::Var(ref mut v) => v.abstract_with(f),
            TypeF::Bit => {}
            TypeF::Array(ref mut elem_ty, ref mut size_expr) => {
                T::as_mut(elem_ty).abstract_level_with(level, f);
                E::as_mut(size_expr).abstract_level_with(level, f);
            }
            TypeF::Cond(Named(_, ref mut ty), ref mut pred_expr) => {
                T::as_mut(ty).abstract_level_with(level, f);
                E::as_mut(pred_expr).abstract_level_with(level, f);
            }
            TypeF::Abs(_, ref mut body_ty) => {
                T::as_mut(body_ty).abstract_level_with(level + 1, f);
            }
            TypeF::Union(ref mut tys) => for ty in tys {
                T::as_mut(ty).abstract_level_with(level, f);
            },
            TypeF::Struct(ref mut fields) => for (i, field) in fields.iter_mut().enumerate() {
                T::as_mut(&mut field.value).abstract_level_with(level + i as u32, f);
            },
            TypeF::App(ref mut fn_ty, ref mut arg_ty) => {
                T::as_mut(fn_ty).abstract_level_with(level, f);
                T::as_mut(arg_ty).abstract_level_with(level, f);
            }
        };
    }

    pub fn abstract_with<F>(&mut self, f: &F)
    where
        F: Fn(&N) -> Option<Named<N, u32>>,
    {
        self.abstract_level_with(0, &f);
    }

    pub fn abstract_name(&mut self, x: &N)
    where
        N: PartialEq + Clone,
    {
        self.abstract_with(&|y| if x == y {
            Some(Named(x.clone(), 0))
        } else {
            None
        });
    }

    fn instantiate_level(&mut self, level: u32, src: &TypeF<N, T, E>)
    where
        N: Clone,
    {
        // Bleh: Running into non-lexical liftetime problems here!
        // Just so you know that I'm not going completely insane....
        // FIXME: ensure that expressions are not bound at the same level
        *self = match *self {
            TypeF::Var(Var::Bound(ref i)) => if *i == level {
                src.clone()
            } else {
                return;
            },
            TypeF::Var(Var::Free(_)) | TypeF::Bit => return,
            TypeF::Array(ref mut elem_ty, _) => {
                T::as_mut(elem_ty).instantiate_level(level, src);
                return;
            }
            TypeF::Cond(Named(_, ref mut ty), _) => {
                T::as_mut(ty).instantiate_level(level, src);
                return;
            }
            TypeF::Union(ref mut tys) => {
                for ty in tys {
                    T::as_mut(ty).instantiate_level(level, src);
                }
                return;
            }
            TypeF::Struct(ref mut fields) => {
                for (i, field) in fields.iter_mut().enumerate() {
                    T::as_mut(&mut field.value).instantiate_level(level + i as u32, src);
                }
                return;
            }
            TypeF::Abs(_, ref mut ty) => {
                T::as_mut(ty).instantiate_level(level + 1, src);
                return;
            }
            TypeF::App(ref mut ty, ref mut arg_ty) => {
                T::as_mut(ty).instantiate_level(level, src);
                T::as_mut(arg_ty).instantiate_level(level, src);
                return;
            }
        };
    }

    pub fn instantiate(&mut self, ty: &TypeF<N, T, E>)
    where
        N: Clone,
    {
        self.instantiate_level(0, ty);
    }

    pub fn repr(&self) -> Result<host::Type<N, E>, ()>
    where
        N: Clone,
    {
        match *self {
            TypeF::Var(ref v) => Ok(host::TypeF::Var(v.clone()).into()),
            TypeF::Bit => Ok(host::TypeF::Bit.into()),
            TypeF::Array(ref elem_ty, ref size_expr) => {
                let elem_repr_ty = Box::new(T::as_ref(elem_ty).repr()?);
                let size_expr = size_expr.clone();

                Ok(host::TypeF::Array(elem_repr_ty, size_expr).into())
            }
            TypeF::Cond(Named(_, ref ty), _) => T::as_ref(ty).repr(),
            TypeF::Union(ref tys) => {
                let repr_tys = tys.iter()
                    .map(T::as_ref)
                    .map(TypeF::repr)
                    .collect::<Result<_, _>>()?;

                Ok(host::TypeF::Union(repr_tys).into())
            }
            TypeF::Struct(ref fields) => {
                let repr_fields = fields
                    .iter()
                    .map(|f| {
                        T::as_ref(&f.value)
                            .repr()
                            .map(|ty| Field::new(f.name.clone(), ty))
                    })
                    .collect::<Result<_, _>>()?;

                Ok(host::TypeF::Struct(repr_fields).into())
            }
            TypeF::Abs(_, _) | TypeF::App(_, _) => Err(()),
        }
    }
}

impl<N, T, E> TypeF<N, T, E>
where
    N: PartialEq,
{
    pub fn lookup_field(&self, name: &N) -> Option<&T> {
        match *self {
            TypeF::Struct(ref fields) => syntax::lookup_field(fields, name),
            _ => None,
        }
    }
}
