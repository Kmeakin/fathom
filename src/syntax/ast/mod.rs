//! The syntax of our data description language

use std::collections::BTreeMap;
use std::rc::Rc;

use name::Name;
use var::ScopeIndex;

pub mod binary;
pub mod host;

/// A field in a struct type
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Field<N, T> {
    pub name: N,
    pub value: T,
}

impl<N, T> Field<N, T> {
    pub fn new<M: Into<N>, U: Into<T>>(name: M, value: U) -> Field<N, T> {
        Field {
            name: name.into(),
            value: value.into(),
        }
    }

    /// Apply the function `f` to the field name and return the wrapped result
    pub fn map_name<M, F: FnMut(N) -> M>(self, mut f: F) -> Field<M, T> {
        Field {
            name: f(self.name),
            value: self.value,
        }
    }

    /// Apply the function `f` to the field value and return the wrapped result
    pub fn map_value<U, F: FnMut(T) -> U>(self, mut f: F) -> Field<N, U> {
        Field {
            name: self.name,
            value: f(self.value),
        }
    }
}

fn lookup_field<'a, N, T>(fields: &'a [Field<N, T>], name: &N) -> Option<&'a T>
where
    N: PartialEq,
{
    fields
        .iter()
        .find(|field| &field.name == name)
        .map(|field| &field.value)
}

/// A type definition
///
/// ```plain
/// Point = {
///     x : u16,
///     y : u16,
/// }
/// ```
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Definition<N> {
    pub name: N,
    pub ty: binary::RcType<N>,
}

impl<N> Definition<N> {
    pub fn new<N1: Into<N>, T1: Into<binary::RcType<N>>>(name: N1, ty: T1) -> Definition<N> {
        Definition {
            name: name.into(),
            ty: ty.into(),
        }
    }
}

pub type Substitutions<N> = BTreeMap<N, binary::Type<N>>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program<N> {
    pub defs: Vec<Definition<N>>,
}

impl<N: Name> Program<N> {
    pub fn new(mut defs: Vec<Definition<N>>) -> Program<N> {
        // We maintain a list of the seen definition names. This will allow us to
        // recover the index of these variables as we abstract later definitions...
        let mut seen_names = Vec::<N>::new();

        for def in &mut defs {
            for (level, name) in seen_names.iter().rev().enumerate() {
                Rc::make_mut(&mut def.ty)
                    .abstract_names_at(&[name.clone()], ScopeIndex(level as u32));
            }

            // Record that the definition has been 'seen'
            seen_names.push(def.name.clone());
        }

        Program { defs }
    }

    pub fn substitute(&mut self, substs: &Substitutions<N>) {
        for def in &mut self.defs {
            Rc::make_mut(&mut def.ty).substitute(substs);
        }
    }
}

pub fn base_defs<N: Name + for<'a> From<&'a str>>() -> Substitutions<N> {
    use syntax::ast::binary::{Type, TypeConst};

    fn prim_array_ty<N: Name>(size: i64, conv_name: &'static str) -> binary::Type<N> {
        use source::Span;
        use syntax::ast::host::Expr;

        let array_ty = Type::array(Span::start(), Type::u8(), Expr::int(Span::start(), size));
        let conv_ty = host::Type::arrow(vec![array_ty.repr()], host::Type::int());

        Type::interp(
            Span::start(),
            array_ty,
            Expr::prim(conv_name, conv_ty),
            host::Type::int(),
        )
    }

    btreemap! {
        // TODO: "true" = Expr::bool(true)
        // TODO: "false" = Expr::bool(false)
        "u8".into() => Type::Const(TypeConst::U8),
        // Little endian primitives
        "u16le".into() => prim_array_ty(2, "from_u16le"),
        "u32le".into() => prim_array_ty(4, "from_u32le"),
        "u64le".into() => prim_array_ty(8, "from_u64le"),
        "i8le".into() => prim_array_ty(8, "from_i8le"),
        "i16le".into() => prim_array_ty(2, "from_i16le"),
        "i32le".into() => prim_array_ty(4, "from_i32le"),
        "i64le".into() => prim_array_ty(8, "from_i64le"),
        "f32le".into() => prim_array_ty(4, "from_f32le"),
        "f64le".into() => prim_array_ty(8, "from_f64le"),
        // Big endian primitives
        "u16be".into() => prim_array_ty(2, "from_u16be"),
        "u32be".into() => prim_array_ty(4, "from_u32be"),
        "u64be".into() => prim_array_ty(8, "from_u64be"),
        "i8be".into() => prim_array_ty(8, "from_i8be"),
        "i16be".into() => prim_array_ty(2, "from_i16be"),
        "i32be".into() => prim_array_ty(4, "from_i32be"),
        "i64be".into() => prim_array_ty(8, "from_i64be"),
        "f32be".into() => prim_array_ty(4, "from_f32be"),
        "f64be".into() => prim_array_ty(8, "from_f64be"),
    }
}
