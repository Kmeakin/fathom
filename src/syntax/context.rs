use im::Vector;
use nameless::{Ignore, Name};
use std::fmt;
use std::rc::Rc;

use syntax::core::{Term, Type};
use syntax::pretty::{self, ToDoc};

/// An entry in the context
#[derive(Debug, Clone, PartialEq)]
pub enum Entry {
    /// A type claim
    Claim(Name, Rc<Type>),
    /// A value definition
    Definition(Name, Rc<Term>),
}

/// A list of binders that have been accumulated during typechecking
///
/// We use a persistent vector internally so that we don't need to deal with the
/// error-prone tedium of dealing with a mutable context when entering and
/// exiting binders.
#[derive(Clone, PartialEq)]
pub struct Context {
    pub entries: Vector<Entry>,
}

impl Context {
    /// Create a new, empty context
    pub fn new() -> Context {
        Context {
            entries: Vector::new(),
        }
    }

    pub fn claim(&self, name: Name, ty: Rc<Type>) -> Context {
        Context {
            entries: self.entries.push_front(Entry::Claim(name, ty)),
        }
    }

    pub fn define(&self, name: Name, term: Rc<Term>) -> Context {
        Context {
            entries: self.entries.push_front(Entry::Definition(name, term)),
        }
    }

    pub fn lookup_claim(&self, name: &Name) -> Option<Rc<Type>> {
        self.entries
            .iter()
            .filter_map(|entry| match *entry {
                Entry::Claim(ref n, ref ty) if n == name => Some(ty.clone()),
                Entry::Claim(_, _) | Entry::Definition(_, _) => None,
            })
            .next()
    }

    pub fn lookup_definition(&self, name: &Name) -> Option<Rc<Term>> {
        self.entries
            .iter()
            .filter_map(|entry| match *entry {
                Entry::Definition(ref n, ref term) if n == name => Some(term.clone()),
                Entry::Definition(_, _) | Entry::Claim(_, _) => None,
            })
            .next()
    }
}

impl Default for Context {
    fn default() -> Context {
        use syntax::core::{Constant, Level, Value};

        let universe0 = Rc::new(Value::Universe(Level(0)));
        let constant = |c| Rc::new(Term::Constant(Ignore::default(), c));
        let constant_val = |c| Rc::new(Value::Constant(c));

        Context::new()
            .claim(Name::user("true"), constant_val(Constant::BoolType))
            .define(Name::user("true"), constant(Constant::Bool(true)))
            .claim(Name::user("false"), constant_val(Constant::BoolType))
            .define(Name::user("false"), constant(Constant::Bool(false)))
            .claim(Name::user("Bool"), universe0.clone())
            .define(Name::user("Bool"), constant(Constant::BoolType))
            .claim(Name::user("String"), universe0.clone())
            .define(Name::user("String"), constant(Constant::StringType))
            .claim(Name::user("Char"), universe0.clone())
            .define(Name::user("Char"), constant(Constant::CharType))
            .claim(Name::user("U8"), universe0.clone())
            .define(Name::user("U8"), constant(Constant::U8Type))
            .claim(Name::user("U16"), universe0.clone())
            .define(Name::user("U16"), constant(Constant::U16Type))
            .claim(Name::user("U32"), universe0.clone())
            .define(Name::user("U32"), constant(Constant::U32Type))
            .claim(Name::user("U64"), universe0.clone())
            .define(Name::user("U64"), constant(Constant::U64Type))
            .claim(Name::user("I8"), universe0.clone())
            .define(Name::user("I8"), constant(Constant::I8Type))
            .claim(Name::user("I16"), universe0.clone())
            .define(Name::user("I16"), constant(Constant::I16Type))
            .claim(Name::user("I32"), universe0.clone())
            .define(Name::user("I32"), constant(Constant::I32Type))
            .claim(Name::user("I64"), universe0.clone())
            .define(Name::user("I64"), constant(Constant::I64Type))
            .claim(Name::user("F32"), universe0.clone())
            .define(Name::user("F32"), constant(Constant::F32Type))
            .claim(Name::user("F64"), universe0.clone())
            .define(Name::user("F64"), constant(Constant::F64Type))
    }
}

impl fmt::Display for Context {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.to_doc().group().render_fmt(pretty::FALLBACK_WIDTH, f)
    }
}

impl fmt::Debug for Context {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        struct FmtContextEntries<'a>(&'a Vector<Entry>);

        impl<'a> fmt::Debug for FmtContextEntries<'a> {
            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
                f.debug_list().entries(self.0).finish()
            }
        }

        f.debug_struct("Context")
            .field("entries", &FmtContextEntries(&self.entries))
            .finish()
    }
}
