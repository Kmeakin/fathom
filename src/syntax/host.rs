//! The syntax of our data description language

use std::fmt;

use source::Span;
use syntax::{self, Field, Name, Named, Var};

#[derive(Copy, Clone, PartialEq, Eq)]
pub enum Const {
    /// A single bit
    Bit(bool),
    /// A boolean constant: eg. `true`, `false`
    Bool(bool),
    /// An integer constant: eg. `0`, `1`, `2`, ...
    Int(i64),
}

impl Const {
    pub fn ty_const_of(self) -> TypeConst {
        match self {
            Const::Bit(_) => TypeConst::Bit,
            Const::Bool(_) => TypeConst::Bool,
            Const::Int(_) => TypeConst::Int,
        }
    }
}

impl fmt::Debug for Const {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            Const::Bit(value) => write!(f, "Bit({:?})", value),
            Const::Bool(value) => write!(f, "Bool({:?})", value),
            Const::Int(value) => write!(f, "Int({:?})", value),
        }
    }
}

/// An unary operator
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum Unop {
    /// Not: eg. `!x`
    Not,
    /// Negation: eg. `-x`
    Neg,
}

/// A binary operator
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum Binop {
    /// Disjunction: eg. `x | y`
    Or,
    /// Conjunction: eg. `x & y`
    And,
    /// Equality: eg. `x == y`
    Eq,
    /// Inequality: eg. `x != y`
    Ne,
    /// Less than or equal: eg. `x <= y`
    Le,
    /// Less than: eg. `x < y`
    Lt,
    /// Greater than: eg. `x > y`
    Gt,
    /// Greater than or equal: eg. `x >= y`
    Ge,
    /// Addition: eg. `x + y`
    Add,
    /// Subtraction: eg. `x - y`
    Sub,
    /// Multiplication: eg. `x * y`
    Mul,
    /// Division: eg. `x / y`
    Div,
}

/// A host expression
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expr<N> {
    /// A constant value
    Const(Span, Const),
    /// Primitive expressions
    Prim(&'static str, Box<Type<N>>),
    /// A variable, referring to an integer that exists in the current
    /// context: eg. `len`, `num_tables`
    Var(Span, Var<N, u32>),
    /// An unary operator expression
    Unop(Span, Unop, Box<Expr<N>>),
    /// A binary operator expression
    Binop(Span, Binop, Box<Expr<N>>, Box<Expr<N>>),
    /// Field projection, eg: `x.field`
    Proj(Span, Box<Expr<N>>, N),
    /// Array index, eg: `x[i]`
    Subscript(Span, Box<Expr<N>>, Box<Expr<N>>),
    /// Abstraction, eg: `\(x : T) -> x`
    Abs(Span, Named<N, Box<Type<N>>>, Box<Expr<N>>),
}

impl<N: Name> Expr<N> {
    /// A bit constant: eg. `0b`, `01`
    pub fn bit(span: Span, value: bool) -> Expr<N> {
        Expr::Const(span, Const::Bit(value))
    }

    /// A boolean constant: eg. `true`, `false`
    pub fn bool(span: Span, value: bool) -> Expr<N> {
        Expr::Const(span, Const::Bool(value))
    }

    /// An integer constant: eg. `0`, `1`, `2`, ...
    pub fn int(span: Span, value: i64) -> Expr<N> {
        Expr::Const(span, Const::Int(value))
    }

    /// Primitive expressions
    pub fn prim<T1: Into<Box<Type<N>>>>(name: &'static str, repr_ty: T1) -> Expr<N> {
        Expr::Prim(name, repr_ty.into())
    }

    /// A free variable, referring to an integer that exists in the current
    /// context: eg. `len`, `num_tables`
    pub fn fvar<N1: Into<N>>(span: Span, x: N1) -> Expr<N> {
        Expr::Var(span, Var::Free(x.into()))
    }

    /// A bound variable
    pub fn bvar<N1: Into<N>>(span: Span, x: N1, i: u32) -> Expr<N> {
        Expr::Var(span, Var::Bound(Named(x.into(), i)))
    }

    /// An unary operator expression
    pub fn unop<E1: Into<Box<Expr<N>>>>(span: Span, op: Unop, x: E1) -> Expr<N> {
        Expr::Unop(span, op, x.into())
    }

    /// A binary operator expression
    pub fn binop<E1, E2>(span: Span, op: Binop, x: E1, y: E2) -> Expr<N>
    where
        E1: Into<Box<Expr<N>>>,
        E2: Into<Box<Expr<N>>>,
    {
        Expr::Binop(span, op, x.into(), y.into())
    }

    /// Field projection, eg: `x.field`
    pub fn proj<E1, N1>(span: Span, expr: E1, field_name: N1) -> Expr<N>
    where
        E1: Into<Box<Expr<N>>>,
        N1: Into<N>,
    {
        Expr::Proj(span, expr.into(), field_name.into())
    }

    /// Array subscript, eg: `x[i]`
    pub fn subscript<E1, E2>(span: Span, expr: E1, index_expr: E2) -> Expr<N>
    where
        E1: Into<Box<Expr<N>>>,
        E2: Into<Box<Expr<N>>>,
    {
        Expr::Subscript(span, expr.into(), index_expr.into())
    }

    /// Abstraction, eg: `\(x : T) -> x`
    pub fn abs<N1, T1, E1>(span: Span, (param_name, param_ty): (N1, T1), body_expr: E1) -> Expr<N>
    where
        N1: Into<N>,
        T1: Into<Box<Type<N>>>,
        E1: Into<Box<Expr<N>>>,
    {
        let param_name = param_name.into();
        let mut body_expr = body_expr.into();
        body_expr.abstract_name(&param_name);
        Expr::Abs(span, Named(param_name, param_ty.into()), body_expr)
    }

    pub fn abstract_name_at(&mut self, name: &N, level: u32) {
        match *self {
            Expr::Var(_, ref mut var) => var.abstract_name_at(name, level),
            Expr::Const(_, _) => {}
            Expr::Prim(_, ref mut repr_ty) => repr_ty.abstract_name_at(name, level),
            Expr::Unop(_, _, ref mut expr) | Expr::Proj(_, ref mut expr, _) => {
                expr.abstract_name_at(name, level);
            }
            Expr::Binop(_, _, ref mut lhs_expr, ref mut rhs_expr) => {
                lhs_expr.abstract_name_at(name, level);
                rhs_expr.abstract_name_at(name, level);
            }
            Expr::Subscript(_, ref mut array_expr, ref mut index_expr) => {
                array_expr.abstract_name_at(name, level);
                index_expr.abstract_name_at(name, level);
            }
            Expr::Abs(_, _, ref mut body_expr) => {
                body_expr.abstract_name_at(name, level + 1);
            }
        }
    }

    pub fn abstract_name(&mut self, name: &N) {
        self.abstract_name_at(name, 0);
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TypeConst {
    /// Bit
    Bit,
    /// Boolean
    Bool,
    /// Integer
    Int,
}

/// A host type
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type<N> {
    /// A type variable: eg. `T`
    Var(Var<N, u32>),
    /// A type constant
    Const(TypeConst),
    /// Arrow type: eg. `T -> U`
    Arrow(Box<Type<N>>, Box<Type<N>>),
    /// An array of the specified type, with a size: eg. `[T; n]`
    Array(Box<Type<N>>, Box<Expr<N>>),
    /// A union of types: eg. `union { T, ... }`
    Union(Vec<Type<N>>),
    /// A struct type, with fields: eg. `struct { field : T, ... }`
    Struct(Vec<Field<N, Type<N>>>),
}

impl<N: Name> Type<N> {
    /// A free type variable: eg. `T`
    pub fn fvar<N1: Into<N>>(x: N1) -> Type<N> {
        Type::Var(Var::Free(x.into()))
    }

    /// A bound type variable
    pub fn bvar<N1: Into<N>>(x: N, i: u32) -> Type<N> {
        Type::Var(Var::Bound(Named(x.into(), i)))
    }

    /// Bit type constant
    pub fn bit() -> Type<N> {
        Type::Const(TypeConst::Bit)
    }

    /// Boolean type constant
    pub fn bool() -> Type<N> {
        Type::Const(TypeConst::Bool)
    }

    /// Integer type constant
    pub fn int() -> Type<N> {
        Type::Const(TypeConst::Int)
    }

    /// Arrow type: eg. `T -> U`
    pub fn arrow<T1, E1>(lhs_ty: T1, rhs_ty: E1) -> Type<N>
    where
        T1: Into<Box<Type<N>>>,
        E1: Into<Box<Type<N>>>,
    {
        Type::Arrow(lhs_ty.into(), rhs_ty.into())
    }

    /// An array of the specified type, with a size: eg. `[T; n]`
    pub fn array<T1, E1>(elem_ty: T1, size_expr: E1) -> Type<N>
    where
        T1: Into<Box<Type<N>>>,
        E1: Into<Box<Expr<N>>>,
    {
        Type::Array(elem_ty.into(), size_expr.into())
    }

    /// A union of types: eg. `union { T, ... }`
    pub fn union(tys: Vec<Type<N>>) -> Type<N> {
        Type::Union(tys)
    }

    /// A struct type, with fields: eg. `struct { field : T, ... }`
    pub fn struct_(mut fields: Vec<Field<N, Type<N>>>) -> Type<N> {
        // We maintain a list of the seen field names. This will allow us to
        // recover the index of these variables as we abstract later fields...
        let mut seen_names = Vec::with_capacity(fields.len());

        for field in &mut fields {
            for (level, name) in seen_names.iter().rev().enumerate() {
                field.value.abstract_name_at(name, level as u32);
            }

            // Record that the field has been 'seen'
            seen_names.push(field.name.clone());
        }

        Type::Struct(fields)
    }

    pub fn lookup_field(&self, name: &N) -> Option<&Type<N>> {
        match *self {
            Type::Struct(ref fields) => syntax::lookup_field(fields, name),
            _ => None,
        }
    }

    pub fn abstract_name_at(&mut self, name: &N, level: u32) {
        match *self {
            Type::Var(ref mut var) => var.abstract_name_at(name, level),
            Type::Const(_) => {}
            Type::Arrow(ref mut lhs_ty, ref mut rhs_ty) => {
                lhs_ty.abstract_name_at(name, level);
                rhs_ty.abstract_name_at(name, level);
            }
            Type::Array(ref mut elem_ty, ref mut size_expr) => {
                elem_ty.abstract_name_at(name, level);
                size_expr.abstract_name_at(name, level);
            }
            Type::Union(ref mut tys) => for ty in tys {
                ty.abstract_name_at(name, level);
            },
            Type::Struct(ref mut fields) => for (i, field) in fields.iter_mut().enumerate() {
                field.value.abstract_name_at(name, level + i as u32);
            },
        }
    }

    pub fn abstract_name(&mut self, name: &N) {
        self.abstract_name_at(name, 0)
    }

    fn instantiate_at(&mut self, level: u32, src: &Type<N>) {
        // FIXME: ensure that expressions are not bound at the same level
        match *self {
            Type::Var(Var::Bound(Named(_, i))) => if i == level {
                *self = src.clone();
            },
            Type::Var(Var::Free(_)) | Type::Const(_) => {}
            Type::Arrow(ref mut lhs_ty, ref mut rhs_ty) => {
                lhs_ty.instantiate_at(level, src);
                rhs_ty.instantiate_at(level, src);
            }
            Type::Array(ref mut elem_ty, _) => {
                elem_ty.instantiate_at(level, src);
            }
            Type::Union(ref mut tys) => for ty in tys {
                ty.instantiate_at(level, src);
            },
            Type::Struct(ref mut fields) => for (i, field) in fields.iter_mut().enumerate() {
                field.value.instantiate_at(level + i as u32, src);
            },
        };
    }

    pub fn instantiate(&mut self, ty: &Type<N>) {
        self.instantiate_at(0, ty);
    }
}
