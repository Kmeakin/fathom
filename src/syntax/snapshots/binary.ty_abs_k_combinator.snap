Abs(
    Named("x", Type),
    Abs(
        Named("y", Type),
        Var(
            Bound(Named("x", 1))
        )
    )
)
